tla-types/src/test/scala/at/forsyte/apalache/tla/typecomp/BuilderTest.scala

```diff
@@ -31,6 +31,157 @@ import scala.collection.immutable.SortedMap
+ *
+ * </ol>
+ *
+ * Assume we want to use the [[runBinary]] template to test our `builder.foo`. Regardless of arity, each template takes
+ * four arguments:
+ *
+ * <ol>
+ *
+ * <li> The `method`. In our case `builder.foo`. [[runBinary]] requires a binary method, other templates require methods
+ * of their corresponding arities. </li>
+ *
+ * <li> A well-typed argument constructor `mkWellTyped`. It should have the type `TParam => T`, i.e., given type
+ * parameters, it should construct a pair of arguments `(x,y)`, such that `build(builder.foo(x,y)` succeeds. In our
+ * case:
+ * {{{
+ *    def mkWellTyped(tparam: TParam): T = {
+ *      val (a, b) = tparam
+ *      (
+ *       builder.name("x", a),
+ *       builder.name("y", FunT1(a, b))
+ *      )
+ *    }
+ * }}}
+ * since the 1st argument must have the type `a`, and the 2nd the type `a -> b`. </li>
+ *
+ * <li> A constructor of ill-typed arguments `mkIllTyped`. It should have the type `TParam => Seq[T]`, i.e., given type
+ * parameters, it should construct a sequence of pairs of arguments `(x,y)`, such that `build(builder.foo(x,y)` produces
+ * a [[TBuilderTypeException]] for each pair. In our case:
+ * {{{
+ *    def mkIllTyped(tparam: TParam): Seq[T] = {
+ *      val (a, b) = tparam
+ *      Seq(
+ *          (
+ *              builder.name("x", InvalidTypeMethods.differentFrom(a)),
+ *              builder.name("y", FunT1(a, b)),
+ *          ),
+ *          (
+ *              builder.name("x", a),
+ *              builder.name("y", InvalidTypeMethods.notApplicative),
+ *          ),
+ *          (
+ *              builder.name("x", a),
+ *              builder.name("y", FunT1(InvalidTypeMethods.differentFrom(a), b)),
+ *          ),
+ *      )
+ *    }
+ * }}}
+ * since there are three ways of violating the type constraints. If `x` has the type `c` and `y` has the type `a -> b`,
+ * where `c /= a`, if `y` doesn't have a function type at all, or if `x` has the type `a` and `y` has the type `c -> b`,
+ * where `c /= a`. Note that the following is _not_ ill-typed:
+ * {{{
+ *    (
+ *        builder.name("x", a),
+ *        builder.name("y", FunT1(a, InvalidTypeMethods.differentFrom(b))),
+ *    )
+ * }}}
+ * since `b` isn't checked against any other argument, unlike `a`, if we swap `b` with some `c` we get arguments with
+ * types `a` and `a -> c`, which is still valid w.r.t. the signature of `foo`, and would return some value of type `c`.
+ * Some operators, for example those that take one argument of any type, have no ill-typed inputs. In those cases,
+ * {{{
+ *    def mkIllTyped(@unused tparam: TParam): Seq[T] = Seq.empty
+ * }}}
+ * </li>
+ *
+ * <li> A validator test `resultIsExpected`. It should have the type `TParam => ResultT => Boolean`, i.e., if `method`
+ * produces a value of type [[TBuilderInternalState]][ResultT] (recall, [[TBuilderInstruction]] is an alias for
+ * [[TBuilderInternalState]][TlaEx]), then `resultIsExpected` is a predicate over values of type `ResultT`
```

Comment:
I think this is more in line of "assumed knowledge" of this HOWTO, (since it doesn't explain `TBuilderInstruction` and the like, that will come in #1912)

Reply:
Sg